Use chunked partition_range vectors for indexed query and in forward_request

bhalevy opened the pull request:
This series uses `utils::chunked_vector<dht::partition_range>` in two places we know could be prone
to large allocations / reactor stalls where we need to allocate a long partition_range vector.

The first case is in `indexed_table_select_statement::find_index_partition_ranges`
where stalls of over 100 ms were seen in the field.

The other case is from `parallelized_select_statement::do_execute` where a cartesian product
of the the query range with lots of vnodes or tablets can result in a long
partition_range vector as well.

perf-simple-query results, using `perf-simple-query -c 1 --duration 10 --random-seed 20240623 --tablets --default-log-level=error`:
```
pre:  98373.34 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42658 insns/op,   21909 cycles/op,        0 errors)
post: 98854.32 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42606 insns/op,   21783 cycles/op,        0 errors)
```

Fixes https://github.com/scylladb/scylladb/issues/18536

Backport to 6.x since the stalls seen in the field were long: over 100ms.

avikivity commented on query_ranges_to_vnodes.cc:
```diff
@@ -107,3 +113,6 @@ void query_ranges_to_vnodes_generator::process_one_range(size_t n, dht::partitio
 }
+
+template class query_ranges_to_vnodes_generator<dht::partition_range_vector>;
+template class query_ranges_to_vnodes_generator<dht::chunked_partition_range_vector>;
```
Seems like bloat.

bhalevy replied:
That's because dht::partition_range_vector isn't converted across the board to chunked_vector.
Just in areas susceptible to long partition_range vectors.

bhalevy replied:
@avikivity I'm not sure I understand your comment.
Would you prefer to convert partition_range_vector to use chunked_vector unconditionally?

bhalevy replied:
> @avikivity I'm not sure I understand your comment. Would you prefer to convert partition_range_vector to use chunked_vector unconditionally?

ping @avikivity

avikivity replied:
I don't think we need chunked_partition_range_vector. Where we have such large vectors, it's due to another, underlying problem.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -600,7 +600,7 @@ indexed_table_select_statement::prepare_command_for_base_query(query_processor&
         query_processor& qp,
-        dht::partition_range_vector&& partition_ranges,
+        dht::chunked_partition_range_vector&& partition_ranges,
         service::query_state& state,
```
Why would this be a large vector? How large would it be?

bhalevy replied:
I don't see where we limit the number of partitions if there are no restrictions on the indexed column, no LIMIT, and the query is unpaged.

bhalevy replied:
I added a unit test that reproduces the large allocation

avikivity replied:
If we don't limit it, won't we OOM?

Using chunked containers is a band-aid if there is no intrinsic limit.

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -364,6 +366,13 @@ chunked_vector<T, max_contiguous_allocation>::chunked_vector(Iterator begin, Ite
+chunked_vector<T, max_contiguous_allocation>::chunked_vector(std::initializer_list<T> x)
+        : chunked_vector() {
+    reserve(x.size());
+    std::move(std::move(x).begin(), std::move(x).end(), std::back_inserter(*this));
```
This doesn't work since initializer_list is a const container.

bhalevy replied:
good point.  will copy instead.

avikivity commented on cql3/restrictions/statement_restrictions.cc:
```diff
@@ -992,7 +992,9 @@ namespace {
 using namespace expr;
 
 /// Computes partition-key ranges from token atoms in ex.
-dht::partition_range_vector partition_ranges_from_token(const expr::expression& ex,
```
> The catesean product of a large query range with

Don't we have a limit on cartesian product sizes?

bhalevy replied:
Good point.
`get_partition_key_ranges` seems safe indeed.

`indexed_table_select_statement::do_execute_base_query` concurrency limit is 4096.
With which, `ranges_to_vnodes` may be asked to return a vector of 4096 partition ranges.
With 120 bytes each, that gets us to 491520 bytes which is a large allocation, but not terribly large.

Cc @raphaelsc

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9906/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9906/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/storage_proxy_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9906/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2231/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10252/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9906/)
#### Failed Tests (4/31000):
- [test_no_amplification_on_scanning_read_2mb](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9906/testReport/junit/read_amplification_test/TestReadAmplification/Tests___dtest_with_topology_changes___test_no_amplification_on_scanning_read_2mb) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_amplification_on_scanning_read_2mb)
- [test_grant_revoke_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9906/testReport/junit/auth_test/TestAuth/Tests___dtest___test_grant_revoke_cleanup) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_grant_revoke_cleanup)
- [test_internode_compression_between_datacenters](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9906/testReport/junit/internode_compression_test/TestInternodeCompression/Tests___dtest___test_internode_compression_between_datacenters) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_internode_compression_between_datacenters)
- [test_no_amplification_on_scanning_read_2mb](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9906/testReport/junit/read_amplification_test/TestReadAmplification/Tests___dtest___test_no_amplification_on_scanning_read_2mb) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_amplification_on_scanning_read_2mb)
#### Build Details:
- Duration: 2 hr 43 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
In v2 [f5f9f21](https://github.com/scylladb/scylladb/commit/f5f9f2129136a7dfb6452cb3e1d05d0105efb9a0):
- rebased
- added `test_large_select_from_index` unit test reproducing #18536

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10973/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10973/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/secondary_index_test
:small_blue_diamond: boost/storage_proxy_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10973/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3090/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11113/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10973/)
#### Failed Tests (5/32335):
- [test_no_amplification_on_scanning_read_2mb](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10973/testReport/junit/read_amplification_test/TestReadAmplification/Tests___dtest_with_topology_changes___test_no_amplification_on_scanning_read_2mb) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_amplification_on_scanning_read_2mb)
- [test_grant_revoke_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10973/testReport/junit/auth_test/TestAuth/Tests___dtest___test_grant_revoke_cleanup) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_grant_revoke_cleanup)
- [test_no_amplification_on_scanning_read_2mb](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10973/testReport/junit/read_amplification_test/TestReadAmplification/Tests___dtest___test_no_amplification_on_scanning_read_2mb) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_amplification_on_scanning_read_2mb)
- [test_scan_paging_missing_limit.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10973/testReport/junit/%28root%29/test_scan/Tests___Unit_Tests___test_scan_paging_missing_limit_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_scan_paging_missing_limit.release.1)
- [alternator.test_scan.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10973/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.release.1)
#### Build Details:
- Duration: 2 hr 52 min
- Builder: spider6.cloudius-systems.com

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1387,19 +1387,19 @@ indexed_table_select_statement::read_posting_list(query_processor& qp,
     const uint64_t limit = get_inner_loop_limit(get_limit(options, _limit), _selection->is_aggregate());
     return read_posting_list(qp, options, limit, state, now, timeout, false).then(utils::result_wrap(
             [this, &options] (::shared_ptr<cql_transport::messages::result_message::rows> rows) {
         auto rs = cql3::untyped_result_set(rows);
```
```c++
// Utility function for reading from the index view (get_index_view()))
// the posting-list for a particular value of the indexed column.
// Remember a secondary index can only be created on a single column.
future<coordinator_result<::shared_ptr<cql_transport::messages::result_message::rows>>>
indexed_table_select_statement::read_posting_list(query_processor& qp,
                  const query_options& options,
                  uint64_t limit,
                  service::query_state& state,
                  gc_clock::time_point now,
                  db::timeout_clock::time_point timeout,
                  bool include_base_clustering_key) const
{
    dht::partition_range_vector partition_ranges = _get_partition_ranges_for_posting_list(options);
    auto partition_slice = _get_partition_slice_for_posting_list(options);

    auto cmd = ::make_lw_shared<query::read_command>(
            _view_schema->id(),
            _view_schema->version(),
            partition_slice,
            qp.proxy().get_max_result_size(partition_slice),
            query::tombstone_limit(qp.proxy().get_tombstone_limit()),
            query::row_limit(limit),
            query::partition_limit(query::max_partitions),
            now,
            tracing::make_trace_info(state.get_trace_state()),
            query_id::create_null_id(),
            query::is_first_page::no,
            options.get_timestamp(state));
```

and

```c++
query::partition_slice indexed_table_select_statement::get_partition_slice_for_local_index_posting_list(const query_options& options) const {
    partition_slice_builder partition_slice_builder{*_view_schema};

    partition_slice_builder.with_ranges(
        _restrictions->get_local_index_clustering_ranges(options, *_view_schema));

    return partition_slice_builder.build();
}
```

It appears we don't set partition_slice::option::allow_short_read, so the size of the result can be arbitrarily large. We probably do add it somewhere though, or we'd see OOMs.

Even if we do set it, the page size of 1MB will translate to lots of small partition keys, most of which will be wasted, since the base table rows will be larger. So we need a smaller read_command::max_result_size (ideally the ratio will be dynamic, but even 1:4 static ratio seems better than what we have now.

Also, query::is_first_page::no? Why?

bhalevy commented:
> If we don't limit it, won't we OOM?
> 
> Using chunked containers is a band-aid if there is no intrinsic limit.

Based on that and the rest of @avikivity's feedback, I think we need to change direction towards applying a limit to restrict the size of the partition_range vectors rather than mitigating the problem with a chunked vector.

So I'll close this PR and hand over the issue to @paszkow and team.

bhalevy commented:
> > If we don't limit it, won't we OOM?
> > Using chunked containers is a band-aid if there is no intrinsic limit.
> 
> Based on that and the rest of @avikivity's feedback, I think we need to change direction towards applying a limit to restrict the size of the partition_range vectors rather than mitigating the problem with a chunked vector.
> 
> So I'll close this PR and hand over the issue to @paszkow and team.

Cc @ndragazis

ndragazis commented:
Thanks @bhalevy, I'm aware of it. I'm trying to figure out what the plan is.

My understanding is that unpaged queries on the index are limited by per-query soft/hard limits (1MiB/100MiB) and a shard-local limit (10% of shard's memory).

For paged queries, the `partition_slice::option::allow_short_read` flag is set implicitly by `query_pager::do_fetch_page()`. Quoting from https://github.com/scylladb/scylladb/commit/1615fe4c5ee:
> All callers should set this already before passing the slice to the
pager, however not all actually do (e.g.
`cql3::indexed_table_select_statement::read_posting_list()`). Instead of
auditing each call site, just make sure this is set in the pager
itself. If someone is creating a pager we can be sure they mean to use
paging.

So, the retrieved partition ranges are bounded in any case. Do we want to enforce an explicit limit through paging, even for unpaged queries?

ndragazis commented:
>Do we want to enforce an explicit limit through paging, even for unpaged queries?

@bhalevy ping.

In principle, the queries on the index view are internal, so they could be paged even if the user query is unpaged.

Code-wise, I'm not sure how this could be done. Currently, querying the index view and the base table are two distinct operations executed sequentially. We construct a vector of partition ranges (or primary keys) which we then feed into `do_execute_base_query()`. One thought would be to replace the vector with a generator-like object that runs paged queries on the index view on demand. We could pick some sane page size there.

On Avi's proposal for a static ratio on `read_command::max_result_size`, it makes sense but I don't see how this is related to the large-vector problem. All three limits in `max_result_size` are configurable via config options (although `do_execute_base_query()` uses a hardcoded value of 1MiB for page size; a bug I suppose?). What if users increase them?

ndragazis commented:
FYI, I sent a new PR: https://github.com/scylladb/scylladb/pull/22254
